explorer/interpreter/interpreter.cpp

```diff
@@ -663,6 +671,49 @@ auto Interpreter::InstantiateWitness(Nonnull<const Witness*> witness)
   return cast<Witness>(value);
 }
 
+auto Interpreter::InstantiateClassWithBase(
+    Nonnull<const StructValue*> init_struct,
+    Nonnull<const NominalClassType*> class_type, SourceLocation source_loc)
+    -> ErrorOr<Nonnull<NominalClassValue*>> {
+  std::vector<NamedValue> struct_values;
+  std::optional<Nonnull<const NominalClassValue*>> base_instance;
+  for (const auto& field : init_struct->elements()) {
+    if (field.name == NominalClassValue::BaseField) {
+      CARBON_CHECK(class_type->base().has_value())
+          << "Invalid 'base' field for class '"
+          << class_type->declaration().name() << "' without base class.";
+      switch (field.value->kind()) {
+        case Value::Kind::NominalClassValue: {
+          base_instance = cast<NominalClassValue>(field.value);
+          break;
+        }
+        case Value::Kind::StructValue: {
+          CARBON_ASSIGN_OR_RETURN(
+              auto base,
+              Convert(field.value, class_type->base().value(), source_loc));
+          base_instance = cast<NominalClassValue>(base);
+          break;
+        }
+        default: {
+          CARBON_FATAL()
+              << "Unhandled base value type while initializing class '"
+              << class_type->declaration().name() << "' from a struct";
+        }
+      }
+    } else {
+      struct_values.push_back(field);
+    }
+  }
+  // Instantiate the `destination_type` to obtain the runtime
+  // type of the object.
+  CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> inst_class,
+                          InstantiateType(class_type, source_loc));
```

Comment:
I think this should be done once, up-front, before you start converting, and as described above, `InstantiateType` should itself recursively instantiate the base class. That way, the `NominalClassValue` you create below will have a type with a correct `base` type rather than one that still contains placeholders.

Reply:
Agree, Done